### ISSUE TITLE
BPE removed during translation

### ISSUE THREAD

**Issue by bricksdont** [joeynmt/joeynmt#91]
**Describe the bug**

I believe that the current inference code internally removes BPE tokens, which in my opinion is unexpected. This happens here:

https://github.com/joeynmt/joeynmt/blob/ea7ac5e1cead107426656938e97d30f6d4c63f9c/joeynmt/prediction.py#L140

since this function for validation is re-used for translation.

**To Reproduce**
Follow all the steps here :)

https://github.com/bricksdont/joeynmt-toy-models

especially:

https://github.com/bricksdont/joeynmt-toy-models/blob/master/scripts/evaluate.sh

where removing BPE from translations does not have any effect.

**Expected behavior**

Translation should not do any postprocessing in my opinion, since JoeyNMT also does not do any preprocessing during training.

Doing BPE postprocessing to show examples from the validation set during training is indeed helpful I think.

**Comment by bastings** [comment]
Seems reasonable to make it return also the not-postprocessed hypotheses. I think it makes sense for Joey to do the post-processing for the sake of computing BLEU, but I can see the point that sometimes you might want the raw output e.g. for analysis. 
Would you like to make a PR?

**Comment by bricksdont** [comment]
Hi Joost,

My perspective is that if Joey does not do any preprocessing, it should also not do any postprocessing.

Will gladly make a PR; by returning also the not-postprocessed hypotheses do you mean that Joey should return both? I was thinking of a switch `postprocess: False` in the config.

**Comment by bastings** [comment]
The question is how BLEU should be calculated by Joey in that case?
Probably not over BPEs (would be misleading)

So I'd say some postprocessing for Bleu/metric computation is necessary.
Is your use case solved if Joey saves raw outputs in addition to
postprocessed ones?

If the issue is consistency then Joey could also do preprocessing :)

**Comment by bricksdont** [comment]
Sorry if I misunderstand, but I think my main point is that validation and test data cannot be handled the same way.

When Joey calculates validation BLEU: yes, it should remove BPE - but actually for a test BLEU it should also undo other preprocessing steps - such as tokenization. Since preprocessing is left to the user, Joey does not know which postprocessing steps to apply.

`sacrebleu.raw_corpus_bleu` should not be used to compute a test BLEU, since there is no standardized, internal tokenization.

**Comment by juliakreutzer** [comment]
Thanks for the discussion @bricksdont and @bastings! 

I agree that the mismatch between pre- and post-processing capabilities is confusing. The 'level' flag basically triggers the post-processing. If you train a model on word-level with 'level=word', it won't do the post-processing. We didn't want to include dynamic pre-processing for memory reasons, but post-processing is cheap and as @bastings BLEU on subwords is not very expressive. 

Regarding BLEU scores on test sets... This is a huge issue itself and papers are rarely transparent on the post-processing of translation outputs and the tokenizers/BLEU scores used (and WMT submissions have a range of hacks such as quotation replacement). So we want to stay out of that mess and hope that Joey NMT users, when they write academic papers and compare against benchmarks, do their research on how model outputs should be post-processed to compute the appropriate BLEU scores. 

The reason for using ´raw_corpus_bleu´ is that we don't know about the tokenizer, and there are some languages where tokenizers don't even exist (e.g. Masakhane benchmarks), so BLEU computation on whatever tokenization is provided in the references seems the most neutral ground. In my opinion, this is still more useful than not reporting any score on the test set.

I propose the following:
- I'll add post-processing and BLEU computation to the FAQs and advice how to get to a score that you can publish.
- If you want to, add raw output logging to the training process or a post-process flag.
- Integrating tokenizers into JoeyNMT is currently in development by a colleague of mine. It requires major restructuring, so I'm not sure when it will be submitted.

What do you think?

**Comment by bricksdont** [comment]
Hi both,

Sorry for the delay!

- Yes, Joey should not do any pre or postprocessing. Just to clarify, I am not suggesting it should. 
- Reporting training progress with `raw_corpus_bleu` on the validation set with BPE removed is completely fine and useful.

Regarding test set evaluation with Joey: in my opinion, evaluation integrated into Joey is not transparent enough at the moment. From a didactic perspective, since we are trying to foster the mindset: "if you publish BLEU scores, it should be case-sensitive on detokenized hypotheses, with standardized tokenization if possible" this should be more transparent.

I would like Joey to return raw output, and then extend the docs to show how to postprocess hypotheses, then compute BLEU with sacreBLEU.

Including maybe how to do it properly for cases like Masakhane systems that do not have tokenizers. I think in this case, users should run sacreBLEU with

    --tokenize none

I see your point that reporting any BLEU score for the test set is more useful than nothing, but this does not mean that the actual translations have to be postprocessed by Joey.

Different take / argument (apologies for this very unstructured response!): If BPE is removed without saving actual model outputs, this prevents users from inspecting model behaviour, e.g. investigating which vocab tokens are produced by the model.

**Proposal:**

Running

    python3 -m joeynmt test

removes BPE if `level=bpe` to report BLEU. However, if there is an `--output_path`, BPE splits are preserved in the file. Docs or BLEU reported on the command line should clarify that this is a score for internal testing, not for publication. As Julia has suggested, docs should include a section on publishable BLEU scores.

Running

    python3 -m joeynmt translate

never does any pre- or postprocessing, which implies it does not remove BPE splits.

Let me know if there are some drawbacks / caveats that I am missing here. If the above is not feasible, I will do a PR with a minimal change for a `postprocess: false` flag.

**Comment by juliakreutzer** [comment]
Hi @bricksdont 
sorry for the delay, forgot to update this thread.

I added some guidelines on evaluation and publishing in the [FAQs](https://joeynmt.readthedocs.io/en/latest/faq.html) in [this commit](https://github.com/joeynmt/joeynmt/commit/86ddee4f02a370a6526d21b75197ed31d356b81a). Please submit a PR if you have suggestions for improvement.

For the time being, we will not modify the evaluation pipeline and the BLEU reporting. However, we will revisit this once tokenizers are integrated.
Would be great if you could go ahead and make the PR with the post-process flag :)

**Comment by bricksdont** [comment]
Hi Julia and Jasmijn!

Finally remembered to do this tiny PR: https://github.com/joeynmt/joeynmt/pull/101

Let me know if there is something I would need to change,

Btw the docs / explanations about which BLEU scores to publish are great!
